### PR TITLE
BPix cabling map and geometry fix

### PR DESCRIPTION
**Pull request by veszpv**
Three bugs found with BPix reconstruction of CRUZET is treated here:
1) the local frame conversion assumed an incorrect orientation for Layer 1 modules on the -Z half of the ladders 
2) a small mistake in the cabling map related to the ROC grouping in Layer 1 modules. 
3) an inverted pattern of flipped and non-flipped ladders were implemented in the BPix geometry for all layers

The PR corrects all bugs from first principle. The fix is mandatory for data-reconstruction, and recommended for future MC production (when data-MC comparison becomes important.)

A new cabling map in Prep called 'SiPixelFedCablingMap_phase1_v7_May2' corrects for 2). The use of this new object and the fix for 1) will invalidate MC generated in earlier CMSSW versions and saved in RAW format. This PR introduces a mechanism which turns on the use of the old local frame conversion if the string "CMSSW_9_0_X" is found in the version name of the cabling map. A new object in Prep called 'SiPixelFedCablingMap_phase1_v6.3_May2' contains this string and the old mapping. Using it will assure backward compatibility.
The fix of 3) will invalidate all earlier MC in GEN-SIM if propagated to the DB.

The following GT-s should be created after merging this PR
- old ideal geometry + cabling v6.3 + old (mis)alignment scenario for compatibility with old MC
- new ideal geometry from xml + cabling v7 + new alignment for collision data reconstruction (and cosmics rereco)
- new ideal geometry from xml + cabling v7 + new misalignment scenario for new MC

Note: the fix in 3) rotated all ladders around by Pi about the direction of global Z, therefore the support structure, the cables, and the cooling tubes moved from one side of the modules to the other for all modules  concerning their arrangement along the global radial direction. The total amount of material budget did not change. Tracking POG was notified that tracking reco material might need to be cross-checked.

### CODE REVIEW

**Comment by franzoni** [comment]
please test

**Comment by ianna** [comment]
Here is Reco geometry comparison with old geometry: new on the left, old on the right. The same DetId is selected. Its global coordinates have changed (in R).
screen shot 2017-05-09 at 17 31 46
screen shot 2017-05-09 at 17 34 03

**Comment by franzoni** [comment]
hello @veszpv thanks;

  may you help with two clarifications:

- does the change in the mapping have an impact on how pixel conditions are assigned to sensitive volumes ?  In other words:  will the update to the cabling map ( SiPixelFedCablingMap_phase1_v7_May2 ) require an update of all other existing pixel-related conditions ? 

- this PR addresses 1) and 3). Are they factored from 2) ? 
   If 2) gets addressed in a separate PR, as you suggest, 
   will the digi2raw and raw2digi for 2017 MC workflow remain consistent ?

@tvami 
please port over the two tags to the prod db, request in 2017+2018 queues and announce; once we'll also have tags for geometry and alignment scenario, we'll proceed with deploying in 92x the third GT (see https://github.com/cms-sw/cmssw/pull/18651#issue-227523146) for release/Mcprduction and to assemble the first GT for studies w/ existing 90-MC. The second GT will actually be an update to the conditions of HTL/exp/prompt.

**Comment by veszpv** [comment]
@franzoni,
- pixel conditions should be fine, 2) intends to change Layer 1 and only the ROC ordering inside a DetId, the FEDchannel-to-DetId mapping remains the same. (Note, fix 3 could lead to a change on the other end of the mapping: the DetId assignment to GeomDetUnits, but this has also been validated, e.g. by @ianna above)
- 2) cannot be validated in a straightforward way without 1) and 3) and vice versa. The only way that could work is by MC starting with GEN-SIM. DIGI-RECO on GEN-SIM will break because of 3), we really have a different geometry.

**Comment by ghellwig** [comment]
@veszpv just to be on the safe side: we can use this geometry now to produce a small private production to validate our fallback asymptotic alignment scenario, right?

We would take your may3 branch together with `SiPixelFedCablingMap_phase1_v7_May2` and your geometry you sent on last Friday, right?

Anything else?

@franzoni should we directly update the realistic alignment scenario with our fallback or use first another hybrid scenario where the pixel is ideal.
The latter might be better to judge the impact of the geometry change, while directly updating might save time.

**Comment by ianna** [comment]
@rovere - do you think recoMaterials.xml should be updated for this geometry change? As expected, there is a difference in BPIX:
tracker_comparisonratio_x_vs_z_vs_r 2017


https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18651/19705/material-budget/comparison/Images/Tracker_ComparisonRatio_x_vs_z_vs_R.png

**Comment by veszpv** [comment]
@ghellwig yes agree, that is all you need.

**Comment by ghellwig** [comment]
@veszpv forgot about the reco material: should we wait for that or is the effect in MC only marginal since we consistently use the same (possibly slightly wrong) material everywhere?

**Comment by ianna** [comment]
@ghellwig - reco material is defined in the scenario, but used only in RECO as it is assigned to reco geometry.

**Comment by ghellwig** [comment]
@ianna that means, we should actually wait for @rovere to see if this gets updated, right?

**Comment by ianna** [comment]
@ghellwig - it's up to you (e.g. POG). In the meantime, I've made a candidate GT you can use 91X_upgrade2017_design_Candidate_2017_05_10_09_23_53

Candidate details:

    Name: 91X_upgrade2017_design_Candidate_2017_05_10_09_23_53
    From GT: 91X_upgrade2017_design_Queue
    Scenario: Upgrade 2017 Simulation
    Workflow: Ideal Conditions
    Release: CMSSW_9_1_X
    Description: Candidate created from the queue: 91X_upgrade2017_design_Queue at: 2017-05-10 09:23:53
    Snapshot Time: 2017-05-10 09:23:53
    Validity: 18446744073709551615
    Your added records:
        Record: GeometryFileRcd , Label: Extended , Tag: XMLFILE_Geometry_92YV1_Extended2017Plan1_mc
        Record: GeometryFileRcd , Label: ExtendedZeroMaterial , Tag: XMLFILE_Geometry_92YV1_Extended2017Plan1ZeroMaterial_mc
        Record: IdealGeometryRecord , Label: - , Tag: TKRECO_Geometry_92YV1
        Record: PTrackerParametersRcd , Label: - , Tag: TKParameters_Geometry_92YV1
        Record: PGeometricDetExtraRcd , Label: - , Tag: TKExtra_Geometry_92YV1

**Comment by ianna** [comment]
+1

**Comment by ghellwig** [comment]
@ianna thank you! This already eases our workflows!

**Comment by tvami** [comment]
@franzoni,
I've uploaded _v7 as SiPixelFedCablingMap_phase1_v7 to Prod and queued to all 91X Upgrade 2017/2018 MC queues, and _v6.3 as SiPixelFedCablingMap_phase1_v6.3 to Prod and  queued to all 90X Upgrade 2017/2018 MC queues. I'll announce to HN now. Should I proceed with updating the data GTs (_hlt) or would you like to do any fast validation before?

**Comment by VinInn** [comment]
One this is done, shall we wait for new LorenzAngles, Templates and clusterShapeFiler calibrations?

**Comment by rovere** [comment]
@ianna I think a re-run of the reconstruction material map is due. I guess we can have the PR in and have a subsequent PR for material and all topics raised by Vincenzo.

**Comment by slava77** [comment]
TTbar13TeV 2017 wf 10024.0 came out with no differences as expected.
Good.
This means that other necessary calibration changes can proceed to be
combined in a bigger effect once this becomes a default.

**Comment by ianna** [comment]
@slava77 - I think, all but 2023 workflows run on geometry from GT.

**Comment by davidlange6** [comment]
Hi @slava77 - 
Right, good indeed.

however, do we understand why this PR gives no change in its current state? I had expected the changes in SiPixelFedCablingMap.cc to create differences until a new GT is included. (obviously my expectations were wrong..)

**Comment by slava77** [comment]
On 5/10/17 6:53 AM, David Lange wrote:
> Hi @slava77 <https://github.com/slava77> -
> Right, good indeed.
> 
> however, do we understand why this PR gives no change in its current
> state? I had expected the changes in SiPixelFedCablingMap.cc to create
> differences until a new GT is included. (obviously my expectations were
> wrong..)

I didn't look at this in depth.

Maybe Viktor can comment.

**Comment by veszpv** [comment]
@davidlange6 @slava77 the new geometry from 3) was not used as pointed out by @ianna  (similarly as 2) was not updated.) The validation starts with gensim as confirmed by @boudoul , all the changes in 1) and 2) should be transparent due to the raw2digi + digi2raw steps.

**Comment by davidlange6** [comment]
Ok, so the statement made in the reco meeting that the GEN-SIM level compatibility of the code would be broken turns out to be incorrect. Great.

**Comment by slava77** [comment]
On 5/10/17 8:14 AM, David Lange wrote:
> Ok, so the statement made in the reco meeting that the GEN-SIM level
> compatibility of the code would be broken turns out to be incorrect. Great.

I have heard otherwise and have even written down so in the minutes on
Friday
https://indico.cern.ch/event/635712/
"The default for MC should preserve compatibility (will change
after/with the relval campaign)"


>

**Comment by davidlange6** [comment]
> On May 10, 2017, at 5:24 PM, Slava Krutelyov <notifications@github.com> wrote:
> 
> On 5/10/17 8:14 AM, David Lange wrote:
> > Ok, so the statement made in the reco meeting that the GEN-SIM level
> > compatibility of the code would be broken turns out to be incorrect. Great.
> 
> I have heard otherwise and have even written down so in the minutes on
> Friday
> https://indico.cern.ch/event/635712/
> "The default for MC should preserve compatibility (will change
> after/with the relval campaign)"

yes, that was the request from me - now, the question is how to active that. I had asked if the gen-sim would still work without further corrections (eg, cabling map), and understood "no". 

Anyway, it makes life easier that there is no compatibility problem. So good.

> 
> 
> >

**Comment by veszpv** [comment]
@davidlange6 @slava77 As I said in the meeting: if you generate MC with the new full package, now with CMSSW_9_2_X (starting with the GEN-SIM step), the code+cabling change is transparent thanks to the digi2raw+raw2digi detour, BUT, compatibility with any GEN-SIM sample generated in earlier releases will break, because we are also changing the geometry.
However, if you use the first GT that I suggested to make: old ideal geometry + cabling v6.3 + old (mis)alignment scenario, compatibility with old MC samples are preserved at all level (even with RAW), but it is not recommended for the 2017 MC campaign, as the geometry in simulation will be different from that in the data. -- Have I answered the question or I misunderstood something?

**Comment by slava77** [comment]
On 5/10/17 9:52 AM, Viktor Veszpremi wrote:
> @davidlange6 <https://github.com/davidlange6> @slava77
> <https://github.com/slava77> As I said in the meeting: if you generate
> MC with the new full package, now with CMSSW_9_2_X (starting with the
> GEN-SIM step), the code+cabling change is transparent thanks to the
> digi2raw+raw2digi detour, BUT, compatibility with any GEN-SIM sample
> generated in earlier releases will break, because we are also changing
> the geometry.
> However, if you use the first GT that I suggested to make: old ideal
> geometry + cabling v6.3 + old (mis)alignment scenario, compatibility
> with old MC samples are preserved at all level (even with RAW), but it
> is not recommended for the 2017 MC campaign, as the geometry in
> simulation will be different from that in the data. -- Have I answered
> the question or I misunderstood something?

This is somewhat different from what I had in mind from the Friday meeting

What I had in mind was:

**Comment by davidlange6** [comment]
but all is ok, as the geometry is conditions driven rather than code driven. So indeed, the default GT can remain as it is until all pieces are in place.

File: CondFormats/SiPixelObjects/src/FrameConversion.cc [hunk 1]
```diff
@@ -8,6 +8,84 @@ using namespace std;
...
+  int slopeRow =0;
+  int slopeCol = 0;
+  int  rowOffset = 0;
+  int  colOffset = 0; 
+ 
+  if (bpix ) { // bpix 
+    
+    if (side==-1 && layer!=1) {  // -Z side: 4 non-flipped modules oriented like 'dddd', except Layer 1
+
+      if (rocIdInDetUnit <8) {
```
Comment (franzoni) [inline CondFormats/SiPixelObjects/src/FrameConversion.cc]: hello @veszpv 
there's plenty of magic numbers in this part of the code
can it be made better ?

**Comment by franzoni** [comment]
@ghellwig 
please for alignment, let's go directly to the realistic scenario both for pixel and strips (i.e. no intermediate step with pixel still ideal); we agreed that the changes stemming from the fix happening here and the deployment of the actual alignment can be still told apart at validation level, if issues were to be present

@veszpv 
@tvami asks me in this thread wether we should deploy the mapping update in data or not already now. To my understanding such update will do little good for unpacking of pixel, as long as  1+2 are not deployed together, which needs 9_1 to go in production.
Can you confirm ?

From this thread I understand that the geometry + mapping update + new alignment (first two available, latter in the making)
are all expected not to enter here, rather in a subsequent GT update PR.
On what time scale are the conditions invoked by @VinInn expecrted to come ?

**Comment by franzoni** [comment]
+1

( other than the magic numbers, ok for AlCa/Db. )

**Comment by veszpv** [comment]
@VinInn since the local directions of the magnetic and electric fields have not changed, there is no need for another LA or template tag.
@rovere so I take for you to be able to verify that the tracking reco material is still ok, you need this PR merged asap?...

**Comment by ghellwig** [comment]
should we actually backport this to 91X?

**Comment by davidlange6** [comment]
no - its for 92x.

**Comment by ghellwig** [comment]
but shouldn't we deploy this as soon as possible to have BPIX data that is actually useful w/o a private rereco?
anyways, I lost a bit track of what 91X is used for in the end...

File: CondFormats/SiPixelObjects/src/FrameConversion.cc (hunk 1 above)
Comment (veszpv) [inline CondFormats/SiPixelObjects/src/FrameConversion.cc]: hello @franzoni, it is dictated by the topology of the hardware, there is no cleaner way of formulating it.

**Comment by mschrode** [comment]
A question from tracker alignment side: will the first collision data (scheduled for week of May22) be taken in 92X? This fix is really needed to have pixel tracks in the express/prompt reconstruction in order to quickly prepare the start-up tracker alignment. If the data taking will happen in 91X, we need a backport.

**Comment by veszpv** [comment]
@franzoni yes, confirmed. The update for the cabling map 2) has little use, without 1) and 3) but it is safe. We should probably append the cabling tag in the data GT as a new IOV before 9_2 is deployed. I guess same applies to also alignment and geometry. Although, updating the geometry would already fix Layer 2-4, and the new alignment already works better for FPix(?). I am not sure what is your protocol in this case.

Concerning MC, my understanding is also such that you will make another PR for the new GT-s containing the correct grouping of the new payloads. Here, the geometry and the cabling are available, the misalignment scenario is missing, and there is no need for updating LA and template.

The Tracking POG maintains the cluster shape filter and the tracking reco material, here I would refer to @VinInn  and @rovere.

**Comment by davidlange6** [comment]
Yes that's why we created a new release cycle dedicated to this fix.....

Cheers,
David


On 12 May 2017, at 18:04, Gregor Mittag <notifications@github.com<mailto:notifications@github.com>> wrote:


but shouldn't we deploy this as soon as possible to have BPIX data that is actually useful w/o a private rereco?
anyways, I lost a bit track of what 91X is used for in the end...

**Comment by davidlange6** [comment]
The plan is 92x

Cheers,
David


On 12 May 2017, at 18:20, Matthias Schroeder <notifications@github.com<mailto:notifications@github.com>> wrote:


A question from tracker alignment side: will the first collision data (scheduled for week of May22) be taken in 92X? This fix is really needed to have pixel tracks in the express/prompt reconstruction in order to quickly prepare the start-up tracker alignment. If the data taking will happen in 91X, we need a backport.

**Comment by ianna** [comment]
@rovere - any news on recoMaterial.xml update? or are you waiting for this PR to be integrated? Thanks.

**Comment by rovere** [comment]
@ianna, in fact, I was waiting for this PR to enter.

**Comment by franzoni** [comment]
after discussion w/ @ghellwig at today's alcadb meeting, 
trying to anticipate what is needed from AlCa/Db to support this process

+ **readiness for data taking** requires:

(i) . code changes of this PR, point 1. from @veszpv 
(ii) . updated mapping, point 2. from @veszpv 
(iii) . reco geometry , see @rovere here above

(ii) and (iii) will be implemented via update of the data GT's, when tags available.

(
@tvami @veszpv : we need to evaluate pro/cons of appending an IOV to [SiPixelFedCablingMap_2009_v1_hlt]( https://cms-conddb.cern.ch/cmsDbBrowser/list/Prod/tags/SiPixelFedCablingMap_2009_v1_hlt ) versus re-doing the tag ; the same conclusion will be adopted for the reco geometry tag; can we skype about this ?
)

It is my understanding that deploying @veszpv 's 1. is undetectable in MC (see comments to the tests earlier in this PR)  ;
and that deploying 1.  w/o deploying 2. and brings no conclusive benefits (and no damage) to 2017 data workflows.
For what concerns data readiness, I'd conclude that only this PR is release bound, and (in absence of 2017 relval tests) all the rest, (ii) and (iii) alike, 
will be handled via append IOV to the HLT/express/Prompt GT's.

+ **readiness for MC** requires:
(i)(ii)(iii) + plus update in the geometry  <= tags all available
reco geometry  + alignment (ETA end-of-this-week) <= tags in the making


Bottom line -  nothing is needed from AlCa/Db to bring this PR to completion;
we'll need to:
 + operate on HLT/express/Prompt GT's as soon as 92x will go to production
 + update GT's in release for MC, in a few days

**Comment by davidlange6** [comment]
I'll merge this PR - but I don't think we should be making a 920 until the reco geometry and things needed for data taking are in place. Thats the whole point of the release after all. Seems we have ~2 days to get that together.

**Comment by fwyzard** [comment]
hi @veszpv 
sorry for the late comment, but I wanted to double check this detail:

> A new cabling map in Prep called 'SiPixelFedCablingMap_phase1_v7_May2' corrects for 2). The use of this new object and the fix for 1) will invalidate MC generated in earlier CMSSW versions and saved in RAW format. This PR introduces a mechanism which turns on the use of the old local frame conversion if the string "CMSSW_9_0_X" is found in the version name of the cabling map. A new object in Prep called 'SiPixelFedCablingMap_phase1_v6.3_May2' contains this string and the old mapping. Using it will assure backward compatibility.

So, if we use SiPixelFedCablingMap_phase1_v6.3_May2 together with CMSSW 9.2.0, we can run over the 8.3.x/9.0.x MC samples in RAW format, without any other special settings (related to the pixel) - correct ?

Can we use this setup together with the recent map of dead modules ?

**Comment by veszpv** [comment]
hi @fwyzard sorry, this comment slipped my attention

Yes, confirmed. All the tags in the GT other than the cabling map and the bad component payload should be the same / compatible as used when the sample was generated.

(I have tested myself a 900 sample with '90X_upgrade2017_realistic_v20' and es_preferring SiPixelFedCablingMap_phase1_v6.3_May2)
